File: python/day2.py
from utils import file_readlines

passwords = file_readlines('../_inputs/day2.txt')


def run_part1():
    print(sum([1 for policy_and_password in passwords if validate_part1(policy_and_password)]))


def run_part2():
    print(sum([1 for policy_and_password in passwords if validate_part2(policy_and_password)]))


def validate_part1(policy_and_password):
    policy, password = policy_and_password.split(':')
    count_range, letter = policy.split(' ')
    min_count, max_count = count_range.split('-')

    min_count = int(min_count)
    max_count = int(max_count)

    count = sum([1 for char in password if char == letter])

    return count >= min_count and count <= max_count


def validate_part2(policy_and_password):
    policy, password = policy_and_password.split(':')
    count_range, letter = policy.split(' ')
    position1, position2 = count_range.split('-')

    position1 = int(position1)
    position2 = int(position2)

    char1 = password[position1]
    char2 = password[position2]

    return sum([1 for char in [char1, char2] if char == letter]) == 1

File: elixir/lib/day2.ex
defmodule Aoc.Day2 do
  @input "day2"

  def run_part1 do
    @input
    |> Aoc.Input.stream_lines()
    |> Enum.filter(fn policy_and_password ->
      [policy, password] = String.split(policy_and_password, ":")
      [range, letter] = String.split(policy, " ")
      [min, max] = range |> String.split("-") |> Enum.map(&String.to_integer/1)

      count =
        password
        |> String.trim()
        |> String.split("")
        |> Enum.count(fn char ->
          char == letter
        end)

      count >= min and count <= max
    end)
    |> Enum.count()
  end

  def run_part2 do
    @input
    |> Aoc.Input.stream_lines()
    |> Enum.filter(fn policy_and_password ->
      [policy, password] = String.split(policy_and_password, ":")
      [range, letter] = String.split(policy, " ")
      [position1, position2] = range |> String.split("-") |> Enum.map(&String.to_integer/1)

      chars = password |> String.trim() |> String.split("") |> Enum.drop(1)

      char1 = Enum.at(chars, position1 - 1)
      char2 = Enum.at(chars, position2 - 1)

      Enum.count([char1, char2], &(&1 == letter)) == 1
    end)
    |> Enum.count()
  end
end
